Domain/SecurityIdentityRetrievalStrategy.php

```diff
@@ -57,8 +58,16 @@ public function getSecurityIdentities(TokenInterface $token)
         }
 
         // add all reachable roles
-        foreach ($this->roleHierarchy->getReachableRoles($token->getRoles()) as $role) {
-            $sids[] = new RoleSecurityIdentity($role);
+        $roles = $this->getRoleNames($token);
+        if (method_exists($this->roleHierarchy, 'getReachableRoleNames')) {
+            foreach ($this->roleHierarchy->getReachableRoleNames($roles) as $role) {
+                $sids[] = new RoleSecurityIdentity($role);
+            }
+        } else {
+            // Symfony < 4.3 BC layer
+            foreach ($this->roleHierarchy->getReachableRoles($roles) as $role) {
```

Comment:
https://github.com/symfony/security-core/blob/4.2/Role/RoleHierarchy.php#L41
getReachableRoles expected array of RoleInterface. Array of string given.
I think `$roles = $this->getRoleNames($token);` need move into condition `if (method_exists($this->roleHierarchy, 'getReachableRoleNames')) {`

Reply:
Can you please send a PR if there's something to fix?

Reply:
Thanks for the quick reply. Yes, of course, I’m doing it right now...

Reply:
@Andrej-in-ua That might trigger another deprecation warning. I think changing the `getRoleNames` method should do the trick.

You can try to change this:

```php
    private function getRoleNames(TokenInterface $token): array
    {
        if (method_exists($token, 'getRoleNames')) {
            return $token->getRoleNames();
        }
        // Symfony < 4.3 BC layer
        return array_map(function (Role $role) {
            return $role->getRole();
        }, $token->getRoles());
    }
```

Into this:

```php
    private function getRoleNames(TokenInterface $token): array
    {
        if (method_exists($token, 'getRoleNames')) {
            return $token->getRoleNames();
        }

        return $token->getRoles();
    }
```

?

Reply:
@XWB In this case, we will rely on the fact that if there is a method `getRoleNames` exists, then there will definitely exists `getReachableRoleNames`. I do not know all the nuances so well. But if you confirm that it is ok, then I do PR.
At least right after I deal with the tests :/

In my project both solution worked.